initial commit of the newest pdf scraper

DanaeBouwer opened the pull request:
**Major Adjustments and restructuring of pdf scraper**

_Summary_
The new scraper takes its input from the masterlist, and loops through a list of article ID's. It directly navigates to the article download page on JSTOR, and automatically downloads it from there. This scraper also integrates the reCAPTCHA solver which was restructured by incorporating cookies to make sure the user stays logged in for the mp3 download and audio recognition to text transformation.

DanaeBouwer committed:
initial commit of the newest pdf scraper

wenkanglu commented on Masterlist_Scraper.py:
```diff
@@ -0,0 +1,192 @@
+    "plugins.always_open_pdf_externally": True, #It will not show PDF directly in chrome
+    "credentials_enable_service": False, # gets rid of password saver popup
+    "profile.password_manager_enabled": False #gets rid of password saver popup
+    })
```
This can all be outside of the loop since none of it changes between iterations.

DanaeBouwer replied:
Cool, done!

wenkanglu commented on Masterlist_Scraper.py:
```diff
@@ -0,0 +1,192 @@
+
+    #     cookies = pickle.load(open("cookies.pkl", "rb"))
+    #     for cookie in cookies:
+    #         driver.add_cookie(cookie)
```
Remove if not needed.

DanaeBouwer replied:
Made the change.

wenkanglu reviewed (changes requested):
General comments:
* We should use a dependency manager like [_Poetry_](https://python-poetry.org/) to keep track of dependencies of the project. It'll also allow whoever else working on this repository to easily download all dependencies.
* Make use of a formatter like [_Black_](https://github.com/psf/black) for consistent styling.
* Follow PEP8 for additional styling and naming conventions.
* Rename Masterlist_Scraper.py to main.py since it is the entry point to the application.
* Remove unnecessary committed files/folders like __pycache__.
* Remove Metadata.csv and Metadata.json and store them in something like Google Drive and point them out in a README. The files are too large and makes pushing and pulling cumbersome.

DanaeBouwer committed:
added feature to rotate user agent

wenkanglu commented on src/connection_controllers/gen_connection_controller.py:
```diff
@@ -5,31 +5,30 @@
-class GenConnectionController(ConnectionController):
 
+class GenConnectionController(ConnectionController):
     def __init__(self, driver: webdriver, host: str):
```
I do not recommend placing any logic in the __init__ function. The __init__ function serves as the "constructor" to our class and should only be used to assign local variables to self. Any additional logic should be placed in their own function and then called upon externally.

wenkanglu commented on src/main.py:
```diff
@@ -26,156 +27,172 @@ def latest_download_file():
         # The t&c's only pop up the first time you start the driver
 
-        while t_c_accepted!=True:
+        while t_c_accepted != True:
```
Can be simplified to `while !t_c_accepted:`

Also, this is potentially infinite if something goes wrong and the t&c's are never accepted. We should add a maximum number of tries (same as any other potentially infinite while loop).

You can do something like `while !t_c_accepted and current_try < MAX_TRIES_FOR_T&C`

In fact, once we do this, we can then log when this maximum number of tries has taken place and we can see where our bottlenecks are during analysis.

DanaeBouwer replied:
The while !t_c_accepted is not working, but adjusted that section to avoid a possible infinite loop.

wenkanglu replied:
Try "while not t_c_accepted". Forgot that this was Python lol

wenkanglu commented on src/main.py:
```diff
@@ -26,156 +27,172 @@ def latest_download_file():
                     break
-        
+
         if restart == True:
```
`if restart`

wenkanglu commented on src/main.py:
```diff
@@ -26,156 +27,172 @@ def latest_download_file():
-        #count = 0
-        while file == url_pending: #and count <= 60:
+        # count = 0
+        while file == url_pending:  # and count <= 60:
```
Why is the count commented out? We want to avoid infinite loops at all cost.

DanaeBouwer replied:
It was for a test, have uncommented it.

wenkanglu commented on src/recaptcha_solver.py:
```diff
@@ -12,17 +12,23 @@
 import requests
 
+
 def delay():
```
Rather add the minimum and maximum time to wait as parameters so you can reuse this delay function wherever you need a random amount of time to wait.

DanaeBouwer replied:
I added a module that calculates the 5 point moving average download time of the user. I replaced the delay function with this and re-use it throughout the programme.

wenkanglu commented on src/recaptcha_solver.py:
```diff
@@ -44,44 +52,47 @@ def check_solved(driver,url,url_pending):
 
     recaptcha_control = True
 
     while recaptcha_control == True:
```
`while recaptcha_control`
Although the variable name does not really sound like a boolean. Maybe think of a different name here.

DanaeBouwer replied:
There are two frames in the reCAPTCHA: a control frame and a challenge frame. I therefore called the respective while loops for each section, recaptcha_control and recaptcha_challenge. Do you still thing the naming convention should be changed?

wenkanglu replied:
Yeah so a boolean should always be asking a question in the form of "Is something true or false". A recaptcha control or challenge being true is vague as to what that really means. For example, if it is regarding whether the control is active, then `is_recaptcha_control_active` is a better name.

DanaeBouwer replied:
Okay, makes sense! Will change the naming appropriately.

wenkanglu commented on src/recaptcha_solver.py:
```diff
@@ -94,56 +105,62 @@ def recaptcha_solver(driver,url,url_pending):
         else:
             recaptcha_challenge = True
-            while recaptcha_challenge==True:
+            while recaptcha_challenge == True:
```
`while recaptcha_challenge`

Change name to be more of a boolean.

DanaeBouwer replied:
Same as the recaptcha_control. Let me know if I should change it.

wenkanglu commented on src/recaptcha_solver.py:
```diff
@@ -213,7 +236,7 @@ def recaptcha_solver(driver,url,url_pending):
                         break
 
-                    if check_solved(driver,url,url_pending)==True:
+                    if check_solved(driver, url, url_pending) == True:
```
` if check_solved(driver, url, url_pending):`

wenkanglu commented on src/temp_storage.py:
```diff
@@ -0,0 +1,19 @@
+
+
+class storage:
+    def getTempStoragePath():
```
Please be consistent with naming. In Python we prefer snake case, e.g., `get_temp_storage_path`

DanaeBouwer replied:
Got it, thanks!

wenkanglu reviewed (changes requested):
Rather minor changes that you can just add with your next commit.

DanaeBouwer committed:
Added additional modules and fixed driver restart login issue

Added internet_speed.py to return user internet speed to determine how long system should sleep between actions
Added user_login to guide the user to login to JSTOR
Fixed the driver restart issue, the user now stays logged in after driver has to restart

Testing:
None

DanaeBouwer committed:
Added additional modules and fixed driver restart login issue

Added internet_speed.py to return user internet speed to determine how long system should sleep between actions
Added user_login to guide the user to login to JSTOR
Fixed the driver restart issue, the user now stays logged in after driver has to restart

Testing:
None
